cobc/codegen.c

```diff
@@ -9681,6 +9681,25 @@ output_report_control (struct cb_report *p, int id, cb_tree ctl, cb_tree nx)
 	if(nx) {
 		output_report_control(p, id, nx, CB_CHAIN(nx));
 	}
+
+	bfound = 0;
+	for(i= p->num_lines-1; i >= 0; i--) {
+		if(p->line_ids[i]->report_control) {
+			struct cb_field *c = cb_code_field (p->line_ids[i]->report_control);
+			if(c == s) {
+				bfound = 1;
+				break;
+			}
+		}
+	}
+	if (!bfound) {
+		cb_warning_x (COBC_WARN_FILLER,
+			CB_TREE(ctl), _("Control field %s is not referenced in report"), s->name);
+		ctl = NULL;
+		p->controls = NULL;
+		return ;
```

Comment:
This won't work if there is a list of more than one controls, and the first one is not used. If you only issue a warning and not an error, your code has to handle correctly such a case.

I would advise to clean up the controls at the beginning of the `output_report_definition` function, so that `p->controls` only contains controls that are used within the report.

Reply:
Just to clarify: `output_control_report` is called on `nx` recursively just before your test, for other controls, so if you set `p->controls = NULL`, you discard all of them. The list of controls is used several times, so doing the test every time would be too complex, the only solution I see is to cleanup the list once at the beginning.